### ISSUE TITLE
How to increment array nested object field?

### ISSUE THREAD

**Issue by zmras** [parse-community/parse-server#6687]
Hi! Let's assume I have the following data:

```
obj = {
  _id: 'someId',
  items: [
    { value: 'a', count: 5 },
    { value: 'b', count: 1 },
  ]
}
```

Please, I want to know how can I use obj.increment() to increment the count field of the first element in items?

I have tried obj.increment('items.0.count') but it doesn't work. I got the following error: Cannot read property '0' of undefined

**Comment by mtrezza** [comment]
What data type is `items`?

**Comment by mtrezza** [comment]
To clarify my previous comment: what you want to do is not documented (i.e. not supported) in Parse Server, but I think you're onto something. It makes sense to support it since it is supported by MongoDB.

I had a look into why this doesn't work and I found several roadblocks:

### 1) Parse JS SDK:

The SDK assumes that a nested key does not contain an array, it fails to traverse the nested path, because `0` is not an object key but an array index in the example above.

https://github.com/parse-community/Parse-SDK-JS/blob/e5094aeb5c5d8b2a396f2171f9cb26a70f94a64b/src/ObjectStateMutations.js#L126-L137

This causes the error: `TypeError: Cannot read property '0' of undefined`.
Workaround is to fetch the object before incrementing and saving.

### 2) Parse JS SDK:

For nested keys the SDK validates the existence of said key, otherwise it does not set the increment operator on that key. Effectively, this does not allow to increment a nested key for an object that has been instantiated with `Parse.Object.createWithoutData()`. I'm not sure why that is only checked for nested keys and what implications it has to change that.

https://github.com/parse-community/Parse-SDK-JS/blob/e5094aeb5c5d8b2a396f2171f9cb26a70f94a64b/src/ParseObject.js#L696-L703

This causes the `increment` instruction to be ignored without throwing an error.
Workaround is to fetch the object before incrementing and saving.
Same root cause as for issue b) in https://github.com/parse-community/parse-server/issues/6455#issue-574022085.

### 3) Parse Server:

SchemaController expects the `Parse.Object` field type to be of `Object` when the key to modify contains a dot:

https://github.com/parse-community/parse-server/blob/6b0efaeffd9fc59002ee87562d0a4105d52d02a3/src/Controllers/SchemaController.js#L1132-L1136

This fails the schema validation because the field in the example above is of type `Array` which causes a type mismatch with the expected type `Object`.

This causes the error:
`ParseError: 111 schema mismatch for Temp.test; expected Array but got Object`

---

Conclusion:
- I would consider roadblocks 1 and 2 to be a bug, regardless of the current issue.
- Roadblock 3 is not a bug but needs to be changed in order to implement the feature of incrementing (or even setting?) values of nested key paths that include arrays.

**Comment by zmras** [comment]
I really thank you for the explaination. I think I will use mongodb directly to increment the field until the feature is considered to be implemented.

**Comment by mtrezza** [comment]
Do you want to reopen this issue to keep track of it?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by zmras** [comment]
Ok! I'll reopen it

**Comment by ruhci28** [comment]
hye, is this issue open .?

**Comment by IngridFuentes** [comment]
Is this issue still open?

**Comment by davimacedo** [comment]
Yes. It is open for a PR. Would you be willed to tackle this one?

**Comment by mtrezza** [comment]
@dplewis Since you are roaming around open issues, if you ever feel like tackling this here, I think it would make many happy.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.2.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/7.2.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.2.0](https://github.com/parse-community/parse-server/releases/tag/7.2.0)
